esp-hal-common/src/dma/mod.rs

```diff
@@ -802,3 +854,313 @@ pub trait DmaTransferRxTx<BR, BT, T>: Drop {
     /// Wait for the transfer to finish.
     fn wait(self) -> (BR, BT, T);
 }
+
+#[cfg(feature = "async")]
+pub(crate) mod asynch {
+    use core::task::Poll;
+
+    use super::*;
+    use crate::macros::interrupt;
+
+    pub struct DmaTxFuture<'a, TX> {
+        pub(crate) tx: &'a mut TX,
+        _a: (),
+    }
+
+    impl<'a, TX> DmaTxFuture<'a, TX>
+    where
+        TX: Tx,
+    {
+        pub fn new(tx: &'a mut TX) -> Self {
+            tx.listen_eof();
+            Self { tx, _a: () }
+        }
+    }
+
+    impl<'a, TX> core::future::Future for DmaTxFuture<'a, TX>
+    where
+        TX: Tx,
+    {
+        type Output = (); // TODO handle DMA errors
+
+        fn poll(
+            self: core::pin::Pin<&mut Self>,
+            cx: &mut core::task::Context<'_>,
+        ) -> Poll<Self::Output> {
+            TX::waker().register(cx.waker());
+            if self.tx.is_listening_eof() {
+                Poll::Pending
+            } else {
+                Poll::Ready(())
+            }
+        }
+    }
+
+    pub struct DmaRxFuture<'a, RX> {
+        pub(crate) rx: &'a mut RX,
+        _a: (),
+    }
+
+    impl<'a, RX> DmaRxFuture<'a, RX>
+    where
+        RX: Rx,
+    {
+        pub fn new(rx: &'a mut RX) -> Self {
+            rx.listen_eof();
+            Self { rx, _a: () }
+        }
+    }
+
+    impl<'a, RX> core::future::Future for DmaRxFuture<'a, RX>
+    where
+        RX: Rx,
+    {
+        type Output = (); // TODO handle DMA errors
+
+        fn poll(
+            self: core::pin::Pin<&mut Self>,
+            cx: &mut core::task::Context<'_>,
+        ) -> Poll<Self::Output> {
+            RX::waker().register(cx.waker());
+            if self.rx.is_listening_eof() {
+                Poll::Pending
+            } else {
+                Poll::Ready(())
+            }
+        }
+    }
+
+    #[cfg(esp32c2)]
+    mod interrupt {
+        use super::*;
+
+        #[interrupt]
+        fn DMA_CH0() {
+            use crate::dma::gdma::{
+                Channel0 as Channel,
+                Channel0RxImpl as ChannelRxImpl,
+                Channel0TxImpl as ChannelTxImpl,
+            };
+
+            if Channel::is_in_done() {
+                Channel::clear_in_interrupts();
```

Comment:
@MabezDev why do you only clear interrupts if the DMA transfer is done? Wouldn't this re-trigger the interrupt handler immediately?